src/plugins/data/common/expressions/types/common.ts

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * This can convert a type into a known Expression string representation of
+ * that type. For example, `TypeToString<Datatable>` will resolve to `'datatable'`.
+ * This allows Expression Functions to continue to specify their type in a
+ * simple string format.
+ */
+export type TypeToString<T> = KnownTypeToString<T> | UnmappedTypeStrings;
+
+/**
+ * Map the type of the generic to a string-based representation of the type.
+ *
+ * If the provided generic is its own type interface, we use the value of
+ * the `type` key as a string literal type for it.
+ */
+// prettier-ignore
+export type KnownTypeToString<T> = 
+  T extends string ? 'string' : 
+  T extends boolean ? 'boolean' : 
+  T extends number ? 'number' :
+  T extends null ? 'null' :
+  T extends { type: string } ? T['type'] :
+  never;
+
+/**
+ * If the type extends a Promise, we still need to return the string representation:
+ *
+ * `someArgument: Promise<boolean | string>` results in `types: ['boolean', 'string']`
+ */
+export type TypeString<T> = KnownTypeToString<UnwrapPromise<T>>;
+
+/**
+ * Types used in Expressions that don't map to a primitive cleanly:
+ *
+ * `date` is typed as a number or string, and represents a date
+ */
+export type UnmappedTypeStrings = 'date' | 'filter';
+
+/**
+ * Utility type: extracts returned type from a Promise.
+ */
+export type UnwrapPromise<T> = T extends Promise<infer P> ? P : T;
```

Comment:
probably we should move it to core, because I already use similar one and found another usage today https://github.com/elastic/kibana/pull/39446#discussion_r298115969

Reply:
Started an issue as discussed: https://github.com/elastic/kibana/issues/39881